### PR TITLE
[PF-110] initial terraform setup for rbs

### PR DESCRIPTION
**Pull request by yonghaoy**

### CODE REVIEW

File: crl-test/api-services.tf [hunk 1]
```diff
@@ -15,5 +15,6 @@ module "enable-services" {
     "storage-component.googleapis.com",
     "bigquery.googleapis.com",
     "cloudbilling.googleapis.com",
+    "compute.googleapis.com",
```
Comment (wchamber) [inline crl-test/api-services.tf]: This should be a different PR?

File: rbs/ip.tf [hunk 2]
```diff
@@ -0,0 +1,7 @@
+resource "google_compute_address" "ingress_ip" {
+  count = var.enable ? 1 : 0
+
+  provider = google.target
+
+  name = "terra-${var.cluster}-${local.owner}-${local.service}-ip"
```
Comment (wchamber) [inline rbs/ip.tf:6]: Can we file/TODO a ticket to follow up on the ingress/firewall we want for RBS? Eventually we need ingress for Rawls (as long as it's not in the same k8s cluster).

File: rbs/sa.tf [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  provider = google.target
+  project  = var.google_project
+  role     = "roles/cloudsql.client"
+  member   = "serviceAccount:${google_service_account.sqlproxy[0].email}"
+}
+
+locals {
+  app_sa_roles = [
+    # Allow creating project.
+    "roles/resourcemanager.projectCreator",
```
Comment (wchamber) [inline rbs/sa.tf]: projectCreator is not a google_project_iam_member role, it belongs on a folder or organization.
We're also gonna need more roles to set up billing accounts, and probably other things like buckets, but maybe best to start with the simplest pieces first, like you have.

File: rbs/variables.tf [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  type        = string
+  description = "Stairway db username"
+  default     = ""
+}
+locals {
+  db_name          = var.db_name == "" ? local.service : var.db_name
+  db_user          = var.db_user == "" ? local.service : var.db_user
+  stairway_db_name = var.stairway_db_name == "" ? "${local.service}-stairway" : var.stairway_db_name
+  stairway_db_user = var.stairway_db_user == "" ? "${local.service}-stairway" : var.stairway_db_user
+}
```
Comment (wchamber) [inline rbs/variables.tf:105]: consider getting rid of the local.service and just inlining these defaults. We're very unlikely to ever change local.service, and all the places where we're using local.service do not need to logically have the same string, so it's not a good dependency indication. Let's keep it simple.

**Review by wchamber — CHANGES_REQUESTED** [review]

File: rbs/variables.tf (hunk 4 above)
Comment (yonghaoy) [reply rbs/variables.tf:105]: Done

File: rbs/ip.tf (hunk 2 above)
Comment (yonghaoy) [reply rbs/ip.tf:6]: We have a Jira ticket RBS service authZ. But I don't think that TODO belongs to here. Here is to generate an ingress_ip address. 
Also, testing tools RBS will need this even Rawls in the same cluster. 
What would be the TODO here?

Comment (wchamber) [reply rbs/ip.tf:6]: sg, no todo.

File: rbs/variables.tf [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  default     = ""
+}
+variable "owner" {
+  type        = string
+  description = "Environment or developer. Defaults to TF workspace name if left blank."
+  default     = ""
+}
+locals {
+  owner   = var.owner == "" ? terraform.workspace : var.owner
+  service = "rbs"
```
Comment (wchamber) [inline rbs/variables.tf:40]: following the same reasoning, can we entirely remove `service`?

File: rbs/sa.tf (hunk 3 above)
Comment (yonghaoy) [reply rbs/sa.tf]: Thanks for pointing the folder permission. For now, I simply put a folder role and make folder_id as input variables. 
In the future, we may want to come back and update here to support multiple folders for different clients. 
And for now to support RBS integration test, I will also enable the service for toolsalpha and create a rbs-testing folder under it.

File: rbs/variables.tf (hunk 5 above)
Comment (yonghaoy) [reply rbs/variables.tf:40]: Done

Comment (yonghaoy) [reply rbs/variables.tf:40]: Actually we need this because other modules are using it, e.g., https://github.com/broadinstitute/terraform-ap-modules/blob/master/crl-janitor/cloudsql.tf#L10

Comment (wchamber) [reply rbs/variables.tf:40]: We don't need the rbs/variables `service` for something in crl-janitor. I think we should replace the `rbs` usage of `service` in cloudsql with the plain string instead of referenceing `local.service`.

Comment (yonghaoy) [reply rbs/variables.tf:40]: Sorry, the crl-janitor link is similar reference RBS will have. 
I feel we should keep the local variable as it a generic template there then we don't need to replace all service name everywhere, then it is easier for us to add modules, and understand other modules? 
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/cloudsql.tf#L10
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/cloudsql.tf#L18
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/cloudsql.tf#L22
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/sa.tf#L7
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/sa.tf#L43
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/sa.tf#L44
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/ip.tf#L6
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/outputs.tf#L43
https://github.com/broadinstitute/terraform-ap-modules/blob/yyu-PF-110/rbs/outputs.tf#L49

Comment (wchamber) [reply rbs/variables.tf:40]: I don't think it's hard to find/replace service names, and I don't know how much copy/pasting we will be doing from/to this module after this initial commit. I think there's some value in being explicit in the names, but not huge deal.

File: rbs/variables.tf [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+variable "stairway_db_user" {
+  type        = string
+  description = "Stairway db username"
+  default     = ""
+}
+locals {
+  db_name          = var.db_name
+  db_user          = var.db_user
+  stairway_db_name = var.stairway_db_name
+  stairway_db_user = var.stairway_db_user
```
Comment (wchamber) [inline rbs/variables.tf]: if we are going to use locals, they should default to something that's not an empty string.

File: terra-env/variables.tf [hunk 7]
```diff
@@ -157,3 +158,11 @@ variable "janitor_google_folder_id" {
   default     = ""
   description = "The folder ID in which Janitor has permission to cleanup resources."
 }
+
+# Terra RBS Vars
+#
+variable "rbs_google_folder_id" {
+  type        = string
+  default     = ""
```
Comment (wchamber) [inline terra-env/variables.tf:166]: I don't know that this should have an empty string default. If we want by default for it to not create folder permissions, null is better. Otherwise, no default will remind us to give it the right folder permissions.

**Review by wchamber — APPROVED** [review]

File: rbs/variables.tf (hunk 5 above)
Comment (yonghaoy) [reply rbs/variables.tf:40]: Yeah, agree. But, there is also value of having common variables, just like our regular code:
```private static final String SERVICE_NAME = "rbs"```
And other place can refer to this something like 
```SERVICE_NAME + "db",  `SERVICE_NAME + "db_name",``` because SERVICE_NAME is used everywhere across the module?

Comment (wchamber) [reply rbs/variables.tf:40]: Common variables are useful for
a) documentation (but here the module name tells us this is all about `rbs`)
b) being able to change something in one place and affect many.  (But we're not likely to rename our service, but it also wouldn't be hard to do later).
c) Making sure different places always use the same value when they need to. (This is not the case for `service_name + "db"` and anything else using the service name. If those values changed independently that wouldn't break anything).
So meh.

Comment (yonghaoy) [reply rbs/variables.tf:40]: Fine with ethier way. But somehow I was implied that McTerra follows similar naming modules, service name as main db, service-stairway as stairway db, hostname is service name, service-{SA role} as SA account naming. And this is the default conversion and falls into c) ?

File: terra-env/variables.tf (hunk 7 above)
Comment (yonghaoy) [reply terra-env/variables.tf:166]: Done

File: rbs/variables.tf (hunk 5 above)
Comment (wchamber) [reply rbs/variables.tf:40]: let's just leave it.

File: rbs/variables.tf (hunk 6 above)
Comment (yonghaoy) [reply rbs/variables.tf]: changed it back to use local.service as default

File: rbs/variables.tf (hunk 5 above)
Comment (gmalkov) [reply rbs/variables.tf:40]: I don't have a super strong opinion on this, slightly leaning towards leaving it as is, but ultimately it's up to y'all. AFAIK changing it won't break anything.

**Review by mikeflinn — APPROVED** [review]
lgtm
